### PR TITLE
[FLINK-33610][runtime] Fix unstable test CoGroupTaskTest.testCancelCoGroupTaskWhileCoGrouping

### PR DESCRIPTION
**Pull request by Jiabao-Sun**
## What is the purpose of the change

[FLINK-33610][runtime] Fix unstable test CoGroupTaskTest.testCancelCoGroupTaskWhileCoGrouping

https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=54741&view=logs&j=675bf62c-8558-587e-2555-dcad13acefb5&t=5878eed3-cc1e-5b12-1ed0-9e7139ce0992&l=7675
```
java.lang.NullPointerException: Cannot invoke "org.apache.flink.runtime.operators.util.CoGroupTaskIterator.close()" because "this.coGroupIterator" is null
at org.apache.flink.runtime.operators.CoGroupDriver.cleanup(CoGroupDriver.java:185)
at org.apache.flink.runtime.operators.testutils.DriverTestBase.testDriverInternal(DriverTestBase.java:270)
at org.apache.flink.runtime.operators.testutils.DriverTestBase.testDriver(DriverTestBase.java:195)
at org.apache.flink.runtime.operators.CoGroupTaskTest.access$200(CoGroupTaskTest.java:45)
at org.apache.flink.runtime.operators.CoGroupTaskTest$3.run(CoGroupTaskTest.java:427)
```

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/operators/testutils/DriverTestBase.java
Comment (XComp) [inline flink-runtime/src/test/java/org/apache/flink/runtime/operators/testutils/DriverTestBase.java:194]:
```suggestion
    protected void testDriver(Driver driver, Class<? extends S> stubClass) throws Exception {
```
I guess, the trick to avoid adding another `\@SuppressWarnings` is to have the generic class being specified in the method signature. :thinking:  WDYT?

File: flink-runtime/src/test/java/org/apache/flink/runtime/operators/testutils/DriverTestBase.java
Comment (XComp) [inline flink-runtime/src/test/java/org/apache/flink/runtime/operators/testutils/DriverTestBase.java:195]:
```suggestion
        testDriver(driver, (S) stubClass.getDeclaredConstructor().newInstance());
```
That's what the IDE is suggesting because `Class'#newInstance()` is deprecated.

File: flink-runtime/src/test/java/org/apache/flink/runtime/operators/testutils/DriverTestBase.java
Comment (XComp) [inline flink-runtime/src/test/java/org/apache/flink/runtime/operators/testutils/DriverTestBase.java:194]:
```suggestion
     /**
     * @deprecated Use {@link #testDriver(Driver, Function)} instead.
     */
    @Deprecated
```
I'm wondering whether we should make this method deprecated and suggest using your newly added method. Passing the class doesn't bring any benefits because we call the default constructor internally anyway. :shrug: 

The only downside of that one would be that all the old test classes are using this method and would have a deprecation warning being add :roll_eyes:

File: flink-runtime/src/test/java/org/apache/flink/runtime/operators/CoGroupTaskTest.java
Comment (XComp) [inline flink-runtime/src/test/java/org/apache/flink/runtime/operators/CoGroupTaskTest.java:342]: I'm not sure whether we're keeping the same test scenario here. The old implementation called `cancel` while `coGroup` was called. The new test implementation is more likely having the cancel call being performed while closing the `RichCoGroupFunction`. WDYT?

**Review by XComp — COMMENTED** [review]
Thanks for your dedication in improving the code base, @Jiabao-Sun. That's much appreciated. I have a few comments. Please check them below.

File: flink-runtime/src/test/java/org/apache/flink/runtime/operators/CoGroupTaskTest.java
Comment (Jiabao-Sun) [reply flink-runtime/src/test/java/org/apache/flink/runtime/operators/CoGroupTaskTest.java]: Thanks @XComp, it makes sense to me.

In the old implementation, the "Thread.sleep" method is used to prevent the `coGroup` operation from ending too early. Therefore, it is highly likely that the "cancel" method is called during the execution of the `coGroup` operation. To maintain consistency with the original scenario, I have adjusted the `OneShotLatch` inside the `coGroup` method.

File: flink-runtime/src/test/java/org/apache/flink/runtime/operators/ReduceTaskTest.java
Comment (XComp) [inline flink-runtime/src/test/java/org/apache/flink/runtime/operators/ReduceTaskTest.java:53]: Same here: Were those changes made accidentally?

File: flink-runtime/src/test/java/org/apache/flink/runtime/operators/ReduceTaskExternalITCase.java
Comment (XComp) [inline flink-runtime/src/test/java/org/apache/flink/runtime/operators/ReduceTaskExternalITCase.java:47]: Why is this change necessary? :thinking:

File: flink-runtime/src/test/java/org/apache/flink/runtime/operators/CoGroupTaskTest.java
Comment (XComp) [inline flink-runtime/src/test/java/org/apache/flink/runtime/operators/CoGroupTaskTest.java:336]:
```suggestion
        final OneShotLatch delayCoGroupProcessingLatch = new OneShotLatch();
```
nit on the naming: Could we add the step that is supposed to be delayed here.

**Review by XComp — COMMENTED** [review]
Thanks for addressing my comments. I have a question about the two code changes that feel like they are added accidentally? Please see my comments below.

File: flink-runtime/src/test/java/org/apache/flink/runtime/operators/ReduceTaskTest.java
Comment (Jiabao-Sun) [reply flink-runtime/src/test/java/org/apache/flink/runtime/operators/ReduceTaskTest.java:53]: This is necessary because we have change the parameter from `Class stubClass` to `Class<? extends S> stubClass`.

The `ReduceTaskTest`'s `MockCombiningReduceStub` implements `GroupReduceFunction`.
If we continue to use `DriverTestBase<RichGroupReduceFunction<Record, Record>>`, we will exceed the specified boundaries.
https://github.com/apache/flink/blob/c61c09e464073fae430cab2dd56bd608f9d275fd/flink-runtime/src/test/java/org/apache/flink/runtime/operators/ReduceTaskTest.java#L316-L318

Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/operators/ReduceTaskTest.java:53]: Ah, I see. I missed that one :+1: Even though, one could argue that instead of fixing the generic you could just switch to the not-deprecated `DriverTestBase#testDriver(Driver, S)` method which you introduced. But I leave it up to you whether you still want to change that. :shrug:

**Review by XComp — APPROVED** [review]
LGTM :+1: Thanks for addressing my comments. I'll wait for CI to go green before merging the PR.

File: flink-runtime/src/test/java/org/apache/flink/runtime/operators/ReduceTaskTest.java
Comment (Jiabao-Sun) [reply flink-runtime/src/test/java/org/apache/flink/runtime/operators/ReduceTaskTest.java:53]: Thanks @XComp for your continuous help. 

Using the newly introduced method `testDriver(Driver, S)` still requires fixing the generics because `S` also restricts the scope of the class. The old method parameter `testDriver(Driver driver, Class stubClass)` does not have any restrictions on the class, so it can be compiled successfully, but it may increase the possibility of errors occurring.

**Pull request merged**
